contracts/system-contracts/hedera-token-service/native-token/HtsFungibleToken.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.4.9 <0.9.0;
+
+import "./interfaces/IHtsFungibleToken.sol";
+import "../IHederaTokenService.sol";
+import "../../HederaResponseCodes.sol";
+import "../KeyHelper.sol";
+
+/**
+ * @title HtsFungibleToken
+ * @dev Abstract contract serving as the base for management contract for a native fungible token implementation, combining IHederaTokenService & ERC20 functionality.
+ */
+abstract contract HtsFungibleToken is IHtsFungibleToken, KeyHelper {
+
+    address constant htsSystemContractAddress = address(0x167);
+    address public htsTokenAddress;
+    
+    event TokenCreated(address tokenAddress, address creatorAddress);
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        int64 _initialTotalSupply,
+        uint8 _decimals
+    ) payable {
+        _createHTSToken(_name, _symbol, _initialTotalSupply, _decimals);
+    }
+
+    function _setupHTSToken(string memory _name, string memory _symbol) internal virtual returns (IHederaTokenService.HederaToken memory tokenInfo) {
+        IHederaTokenService.TokenKey[] memory keys = new IHederaTokenService.TokenKey[](0);
+        IHederaTokenService.Expiry memory expiry = IHederaTokenService.Expiry(0, address(this), 8000000);
+        tokenInfo = IHederaTokenService.HederaToken(
+            _name, _symbol, address(this), "HtsFungibleToken creation", true, 5000, false, keys, expiry
+        );
+    }
+
+    function _createHTSToken(string memory _name, string memory _symbol, int64 _initialTotalSupply, uint8 _decimals) internal virtual {
+        IHederaTokenService.HederaToken memory tokenInfo = _setupHTSToken(_name, _symbol);
+
+        (bool success, bytes memory result) = htsSystemContractAddress.call{value : msg.value}(
+            abi.encodeWithSelector(IHederaTokenService.createFungibleToken.selector,
+            tokenInfo, _initialTotalSupply, _decimals));
+
+        (int responseCode, address tokenAddress) = success ? abi.decode(result, (int32, address)) : (HederaResponseCodes.UNKNOWN, address(0));
+
+        require(responseCode == HederaResponseCodes.SUCCESS, "Failed to create HTS token");
```

Comment:
this pattern masks the `responseCode` in case of error, making it difficult to debug. Should we introduce a custom error, such as `HtsError(int)`, so at least the caller can see the error code?